File: csharp/Patterns/Strategy.cs
using System;

namespace DesignPatterns.Patterns;

public static class Strategy
{
    private interface IPaymentStrategy
    {
        string Pay(decimal amount);
    }

    private class CardPayment : IPaymentStrategy
    {
        public string Pay(decimal amount) => $"Paid ${amount:0.00} with card";
    }

    private class CashPayment : IPaymentStrategy
    {
        public string Pay(decimal amount) => $"Paid ${amount:0.00} with cash";
    }

    private class Checkout
    {
        private readonly IPaymentStrategy _strategy;

        public Checkout(IPaymentStrategy strategy)
        {
            _strategy = strategy;
        }

        public string Complete(decimal amount) => _strategy.Pay(amount);
    }

    public static void Run()
    {
        Console.WriteLine("\nStrategy");
        Console.WriteLine(new Checkout(new CardPayment()).Complete(15m));
        Console.WriteLine(new Checkout(new CashPayment()).Complete(15m));
    }
}

File: python/patterns/strategy.py
from abc import ABC, abstractmethod


class PaymentStrategy(ABC):
    @abstractmethod
    def pay(self, amount: float) -> str:
        raise NotImplementedError


class CardPayment(PaymentStrategy):
    def pay(self, amount: float) -> str:
        return f"Paid ${amount:.2f} with card"


class CashPayment(PaymentStrategy):
    def pay(self, amount: float) -> str:
        return f"Paid ${amount:.2f} with cash"


class Checkout:
    def __init__(self, strategy: PaymentStrategy) -> None:
        self.strategy = strategy

    def complete(self, amount: float) -> str:
        return self.strategy.pay(amount)


def run() -> None:
    print("\nStrategy")
    print(Checkout(CardPayment()).complete(15.0))
    print(Checkout(CashPayment()).complete(15.0))
